android/capacitor/src/main/java/com/getcapacitor/Bridge.java

```diff
@@ -713,13 +758,136 @@ public void onRequestPermissionsResult(int requestCode, String[] permissions, in
+                }
+            }
         } else {
             // Call deprecated method if using deprecated NativePlugin annotation
             plugin.getInstance().handleRequestPermissionsResult(requestCode, permissions, grantResults);
         }
     }
 
+    /**
+     * Saves permission states and rejects if permissions were not correctly defined in
+     * the AndroidManifest.xml file.
+     *
+     * Plugins overriding {@link #onRequestPermissionsResult(int, String[], int[])} should call
+     * this method to save permission states correctly.
+     *
+     * @param plugin
+     * @param savedCall
+     * @param permissions
+     * @param grantResults
+     * @return true if permissions were saved and defined correctly, false if not
+     */
+    protected boolean validatePermissions(Plugin plugin, PluginCall savedCall, String[] permissions, int[] grantResults) {
+        SharedPreferences prefs = getContext().getSharedPreferences(PERMISSION_PREFS_NAME, Activity.MODE_PRIVATE);
+
+        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            // Permission granted. If previously denied, remove cached state
+            for (String permission : permissions) {
+                String state = prefs.getString(permission, null);
+
+                if (state != null) {
+                    SharedPreferences.Editor editor = prefs.edit();
+                    editor.remove(permission);
+                    editor.apply();
+                }
+            }
+        } else {
+            for (String permission : permissions) {
+                SharedPreferences.Editor editor = prefs.edit();
+
+                if (ActivityCompat.shouldShowRequestPermissionRationale(getActivity(), permission)) {
+                    // Permission denied, can prompt again with rationale
+                    editor.putString(permission, "prompt-with-rationale");
+                } else {
+                    // Permission denied permanently, store this state for future reference
+                    editor.putString(permission, "denied");
+                }
+
+                editor.apply();
+            }
+        }
+
+        if (!plugin.hasDefinedPermissions(permissions)) {
+            StringBuilder builder = new StringBuilder();
+            builder.append("Missing the following permissions in AndroidManifest.xml:\n");
+            String[] missing = PermissionHelper.getUndefinedPermissions(getContext(), permissions);
+            for (String perm : missing) {
+                builder.append(perm + "\n");
+            }
+            savedCall.reject(builder.toString());
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Helper to check all permissions and see the current states of each permission.
+     *
+     * @since 3.0.0
+     * @return A mapping of permissions to the associated granted status.
+     */
+    public JSObject getPermissionStates(Plugin plugin) {
```

Comment:
We should make this protected. If the API needs to be exposed to the plugin developer, we can make a helper in the Plugin class that just does `this.bridge.getPermissionStates(this)` (so the usage would be `this.getPermissionStates()`, much cleaner).

Reply:
I will do that. We have the new code in permission states that indicates if its a `prompt-with-rational` or hard deny, and I think that could be useful for plugin authors to know from this helper.